File: src/loom/indexer/adapters/rust.py
"""Rust language adapter for Loom."""

import logging
from pathlib import Path

import tree_sitter_rust as tsrust
from tree_sitter import Language, Node, Parser

from loom.store.models import ParsedEdge, Symbol

log = logging.getLogger(__name__)

RUST_LANGUAGE = Language(tsrust.language())

_RUST_EXTENSIONS: frozenset[str] = frozenset({".rs"})

_RUST_EXCLUDED_DIRS: frozenset[str] = frozenset({"target"})


class RustAdapter:
    """LanguageAdapter for Rust source files."""

    extensions: frozenset[str] = _RUST_EXTENSIONS
    language_name: str = "rust"
    excluded_dirs: frozenset[str] = _RUST_EXCLUDED_DIRS

    def parse(
        self,
        source: bytes,
        file_path: str,
    ) -> tuple[list[Symbol], list[ParsedEdge]]:
        """Parse Rust source bytes and extract symbols and edges."""
        suffix = Path(file_path).suffix
        if suffix not in _RUST_EXTENSIONS:
            return [], []

        parser = Parser(RUST_LANGUAGE)
        tree = parser.parse(source)
        root = tree.root_node

        symbols: list[Symbol] = []
        edges: list[ParsedEdge] = []

        _walk_node(root, source, file_path, symbols, edges, impl_type=None)

        log.debug(
            "Parsed %s: %d symbols, %d edges",
            Path(file_path).name,
            len(symbols),
            len(edges),
        )
        return symbols, edges

    def resolve_module_path(
        self,
        import_path: str,
        source_file: str,
        known_files: set[str],
    ) -> str:
        """Resolve a Rust use/mod path to an actual indexed file.

        Handles crate::, super::, and bare mod names.
        Returns import_path unchanged if no match found.
        """
        if import_path in known_files:
            return import_path

        source_dir = str(Path(source_file).parent)

        # Strip crate:: prefix → resolve from project root
        if import_path.startswith("crate::"):
            remainder = import_path[len("crate::") :]
            path_part = remainder.replace("::", "/")
            for f in known_files:
                # Match files that end with /path.rs, /path/mod.rs, or equal path.rs
                if f.endswith(f"/{path_part}.rs") or f == path_part + ".rs":
                    return f
                if f.endswith(f"/{path_part}/mod.rs") or f == path_part + "/mod.rs":
                    return f

        # Strip super:: prefix → resolve from parent directory
        elif import_path.startswith("super::"):
            remainder = import_path[len("super::") :]
            path_part = remainder.replace("::", "/")
            parent_dir = str(Path(source_dir).parent)
            for candidate in (
                f"{parent_dir}/{path_part}.rs",
                f"{parent_dir}/{path_part}/mod.rs",
            ):
                if candidate in known_files:
                    return candidate

        else:
            # mod foo style — try source file's directory
            path_part = import_path.replace("::", "/")
            for candidate in (
                f"{source_dir}/{path_part}.rs",
                f"{source_dir}/{path_part}/mod.rs",
                path_part + ".rs",
                path_part + "/mod.rs",
            ):
                if candidate in known_files:
                    return candidate

        return import_path


# ── Module-private helpers ───────────────────────────────────────────────────


def _get_text(node: Node, source: bytes) -> str:
    return source[node.start_byte : node.end_byte].decode("utf-8", errors="replace")


def _get_context(source: bytes, node: Node, max_lines: int = 10) -> str:
    start = node.start_point[0]
    end = min(node.end_point[0] + 1, start + max_lines)
    lines = source.split(b"\n")[start:end]
    return b"\n".join(lines).decode("utf-8", errors="replace")


def _child_by_type(node: Node, type_name: str) -> Node | None:
    for child in node.children:
        if child.type == type_name:
            return child
    return None


def _children_by_type(node: Node, type_name: str) -> list[Node]:
    return [c for c in node.children if c.type == type_name]


def _walk_node(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    impl_type: str | None,
) -> None:
    ntype = node.type

    if ntype == "use_declaration":
        _handle_use_declaration(node, source, file_path, edges)
        return

    if ntype == "function_item":
        _handle_function_item(node, source, file_path, symbols, edges, impl_type)
        return

    if ntype == "function_signature_item":
        # Trait method stub — extract as method if in impl_type context
        if impl_type:
            _handle_function_signature(node, source, file_path, symbols, impl_type)
        return

    if ntype == "struct_item":
        _handle_struct_item(node, source, file_path, symbols)
        return

    if ntype == "enum_item":
        _handle_enum_item(node, source, file_path, symbols)
        return

    if ntype == "trait_item":
        _handle_trait_item(node, source, file_path, symbols, edges)
        return

    if ntype == "impl_item":
        _handle_impl_item(node, source, file_path, symbols, edges)
        return

    if ntype == "type_item":
        _handle_type_item(node, source, file_path, symbols)
        return

    if ntype == "const_item":
        _handle_const_static_item(node, source, file_path, symbols, "identifier")
        return

    if ntype == "static_item":
        _handle_const_static_item(node, source, file_path, symbols, "identifier")
        return

    if ntype == "macro_definition":
        _handle_macro_definition(node, source, file_path, symbols)
        return

    for child in node.children:
        _walk_node(child, source, file_path, symbols, edges, impl_type)


def _handle_use_declaration(
    node: Node,
    source: bytes,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Handle `use` statements. Skip glob imports (`use ...*`)."""
    for child in node.children:
        if child.type == "use_wildcard":
            log.warning("Skipping glob use import in %s", file_path)
            return
        if child.type == "scoped_identifier":
            path_text = _get_text(child, source)
            # Extract last segment as symbol name
            last_part = path_text.split("::")[-1]
            edges.append(
                ParsedEdge(
                    source_name=last_part,
                    target_name=last_part,
                    target_file=path_text,
                    relationship="imports",
                )
            )
        elif child.type == "scoped_use_list":
            # e.g., use foo::{A, B}
            _handle_scoped_use_list(child, source, file_path, edges)
        elif child.type == "use_list":
            _handle_use_list(child, source, file_path, edges)
        elif child.type == "use_as_clause":
            # use X as Y — emit edge for the original path X
            orig_node = child.children[0] if child.children else None
            if orig_node:
                orig = _get_text(orig_node, source)
                last_part = orig.split("::")[-1]
                edges.append(
                    ParsedEdge(
                        source_name=last_part,
                        target_name=last_part,
                        target_file=orig,
                        relationship="imports",
                    )
                )
        elif child.type == "identifier":
            name = _get_text(child, source)
            edges.append(
                ParsedEdge(
                    source_name=name,
                    target_name=name,
                    target_file=name,
                    relationship="imports",
                )
            )


def _handle_scoped_use_list(
    node: Node,
    source: bytes,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Handle `use foo::{A, B, C}`."""
    # Find the prefix (scoped_identifier before the use_list)
    prefix = ""
    for child in node.children:
        if child.type in ("scoped_identifier", "identifier"):
            prefix = _get_text(child, source)
        elif child.type == "use_list":
            _handle_use_list(child, source, file_path, edges, prefix=prefix)


def _handle_use_list(
    node: Node,
    source: bytes,
    file_path: str,
    edges: list[ParsedEdge],
    prefix: str = "",
) -> None:
    """Handle `{A, B, C}` in a use statement."""
    for child in node.children:
        if child.type == "use_wildcard":
            log.warning("Skipping glob use import in %s", file_path)
        elif child.type == "identifier":
            name = _get_text(child, source)
            full_path = f"{prefix}::{name}" if prefix else name
            edges.append(
                ParsedEdge(
                    source_name=name,
                    target_name=name,
                    target_file=full_path,
                    relationship="imports",
                )
            )
        elif child.type == "scoped_identifier":
            path_text = _get_text(child, source)
            last_part = path_text.split("::")[-1]
            full_path = f"{prefix}::{path_text}" if prefix else path_text
            edges.append(
                ParsedEdge(
                    source_name=last_part,
                    target_name=last_part,
                    target_file=full_path,
                    relationship="imports",
                )
            )
        elif child.type == "use_as_clause":
            # use X as Y
            orig_node = child.children[0] if child.children else None
            if orig_node:
                orig = _get_text(orig_node, source)
                last_part = orig.split("::")[-1]
                full_path = f"{prefix}::{orig}" if prefix else orig
                edges.append(
                    ParsedEdge(
                        source_name=last_part,
                        target_name=last_part,
                        target_file=full_path,
                        relationship="imports",
                    )
                )


def _handle_function_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
    impl_type: str | None,
) -> None:
    name_node = _child_by_type(node, "identifier")
    if name_node is None:
        return
    raw_name = _get_text(name_node, source)

    if impl_type:
        full_name = f"{impl_type}.{raw_name}"
        kind = "method"
    else:
        full_name = raw_name
        kind = "function"

    symbols.append(
        Symbol(
            name=full_name,
            kind=kind,
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )

    body = _child_by_type(node, "block")
    if body:
        _extract_calls(body, source, full_name, file_path, edges)


def _handle_function_signature(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    impl_type: str,
) -> None:
    """Extract trait method stub as a method symbol."""
    name_node = _child_by_type(node, "identifier")
    if name_node is None:
        return
    raw_name = _get_text(name_node, source)
    full_name = f"{impl_type}.{raw_name}"
    symbols.append(
        Symbol(
            name=full_name,
            kind="method",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )


def _handle_struct_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
) -> None:
    name_node = _child_by_type(node, "type_identifier")
    if name_node is None:
        return
    symbols.append(
        Symbol(
            name=_get_text(name_node, source),
            kind="class",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )


def _handle_enum_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
) -> None:
    name_node = _child_by_type(node, "type_identifier")
    if name_node is None:
        return
    enum_name = _get_text(name_node, source)

    symbols.append(
        Symbol(
            name=enum_name,
            kind="class",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )

    # Extract enum variants
    variant_list = _child_by_type(node, "enum_variant_list")
    if variant_list:
        for child in variant_list.children:
            if child.type == "enum_variant":
                id_node = _child_by_type(child, "identifier")
                if id_node:
                    variant_name = _get_text(id_node, source)
                    symbols.append(
                        Symbol(
                            name=f"{enum_name}.{variant_name}",
                            kind="variable",
                            file=file_path,
                            line=child.start_point[0] + 1,
                            end_line=child.end_point[0] + 1,
                            language="rust",
                            context=_get_context(source, child),
                        )
                    )


def _handle_trait_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
) -> None:
    name_node = _child_by_type(node, "type_identifier")
    if name_node is None:
        return
    trait_name = _get_text(name_node, source)

    symbols.append(
        Symbol(
            name=trait_name,
            kind="class",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )

    # Extract method stubs from the trait body
    body = _child_by_type(node, "declaration_list")
    if body:
        for child in body.children:
            if child.type == "function_signature_item":
                _handle_function_signature(child, source, file_path, symbols, trait_name)
            elif child.type == "function_item":
                _handle_function_item(child, source, file_path, symbols, edges, trait_name)


def _handle_impl_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    edges: list[ParsedEdge],
) -> None:
    """Handle impl blocks: `impl StructName {}` and `impl Trait for Struct {}`."""
    # Collect type_identifier children and check for 'for' keyword
    type_ids: list[Node] = []
    has_for = False

    for child in node.children:
        if child.type == "type_identifier":
            type_ids.append(child)
        elif child.type == "for":
            has_for = True
        elif child.type in ("generic_type",):
            # e.g., impl<T> Trait<T> for Struct<T>
            inner = _child_by_type(child, "type_identifier")
            if inner:
                type_ids.append(inner)

    if not type_ids:
        return

    if has_for and len(type_ids) >= 2:
        # impl TraitName for StructName
        trait_name = _get_text(type_ids[0], source)
        struct_name = _get_text(type_ids[-1], source)
        # Emit implements edges
        edges.append(
            ParsedEdge(
                source_name=struct_name,
                target_name=trait_name,
                target_file=None,
                relationship="implements",
            )
        )
        edges.append(
            ParsedEdge(
                source_name=trait_name,
                target_name=struct_name,
                target_file=None,
                relationship="implemented_by",
            )
        )
        impl_type = struct_name
    else:
        struct_name = _get_text(type_ids[-1], source)
        impl_type = struct_name

    # Recurse into the declaration_list with impl_type set
    body = _child_by_type(node, "declaration_list")
    if body:
        for child in body.children:
            _walk_node(child, source, file_path, symbols, edges, impl_type)


def _handle_type_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
) -> None:
    name_node = _child_by_type(node, "type_identifier")
    if name_node is None:
        return
    symbols.append(
        Symbol(
            name=_get_text(name_node, source),
            kind="variable",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )


def _handle_const_static_item(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
    name_type: str,
) -> None:
    name_node = _child_by_type(node, name_type)
    if name_node is None:
        return
    symbols.append(
        Symbol(
            name=_get_text(name_node, source),
            kind="variable",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )


def _handle_macro_definition(
    node: Node,
    source: bytes,
    file_path: str,
    symbols: list[Symbol],
) -> None:
    name_node = _child_by_type(node, "identifier")
    if name_node is None:
        return
    symbols.append(
        Symbol(
            name=_get_text(name_node, source),
            kind="macro",
            file=file_path,
            line=node.start_point[0] + 1,
            end_line=node.end_point[0] + 1,
            language="rust",
            context=_get_context(source, node),
        )
    )


def _extract_calls(
    node: Node,
    source: bytes,
    caller_name: str,
    file_path: str,
    edges: list[ParsedEdge],
) -> None:
    """Recursively extract call edges from a Rust function/method body."""
    if node.type == "call_expression":
        func_node = node.children[0] if node.children else None
        if func_node and func_node.type in ("identifier", "scoped_identifier", "field_expression"):
            callee = _get_text(func_node, source)
            if callee != caller_name:
                edges.append(
                    ParsedEdge(
                        source_name=caller_name,
                        target_name=callee,
                        target_file=None,
                        relationship="calls",
                    )
                )

    for child in node.children:
        _extract_calls(child, source, caller_name, file_path, edges)

File: crates/loom-core/src/parsers/rust.rs
use std::{collections::BTreeSet, path::Path};

use tree_sitter::Node;

use crate::{
    parsers::{
        tree_sitter_utils::{
            child_by_field, edge, first_named_child_of_kind, language_parse_result, path_parent,
            push_call_edges, qualified, resolve_direct_or_candidates, symbol, text, walk_preorder,
            Scope,
        },
        LanguageAdapter, ParseResult,
    },
    Result,
};

pub struct RustAdapter;

impl LanguageAdapter for RustAdapter {
    fn extensions(&self) -> &'static [&'static str] {
        &[".rs"]
    }

    fn language_name(&self) -> &'static str {
        "rust"
    }

    fn excluded_dirs(&self) -> &'static [&'static str] {
        &["target"]
    }

    fn parse(&self, source: &[u8], file_path: &str) -> Result<ParseResult> {
        if !file_path.ends_with(".rs") {
            return Ok(ParseResult::default());
        }
        language_parse_result(
            source,
            file_path,
            "rust",
            tree_sitter_rust::LANGUAGE.into(),
            |root, result| walk_node(root, source, file_path, result, Scope::TOP),
        )
    }

    fn resolve_module_path(
        &self,
        import_path: &str,
        source_file: &str,
        known_files: &BTreeSet<String>,
    ) -> String {
        if known_files.contains(import_path) {
            return import_path.to_string();
        }
        let source_dir = path_parent(source_file);
        let source_dir_string = source_dir.to_string_lossy();
        if let Some(remainder) = import_path.strip_prefix("crate::") {
            let path = remainder.replace("::", "/");
            let candidates = [format!("{path}.rs"), format!("{path}/mod.rs")];
            return known_files
                .iter()
                .find(|file| {
                    candidates.iter().any(|candidate| {
                        *file == candidate || file.ends_with(&format!("/{candidate}"))
                    })
                })
                .cloned()
                .unwrap_or_else(|| import_path.to_string());
        }
        if let Some(remainder) = import_path.strip_prefix("super::") {
            let parent = source_dir.parent().unwrap_or_else(|| Path::new(""));
            let path = remainder.replace("::", "/");
            return resolve_direct_or_candidates(
                import_path,
                known_files,
                [
                    parent
                        .join(format!("{path}.rs"))
                        .to_string_lossy()
                        .into_owned(),
                    parent
                        .join(format!("{path}/mod.rs"))
                        .to_string_lossy()
                        .into_owned(),
                ],
            );
        }
        let path = import_path.replace("::", "/");
        resolve_direct_or_candidates(
            import_path,
            known_files,
            [
                format!("{source_dir_string}/{path}.rs"),
                format!("{source_dir_string}/{path}/mod.rs"),
                format!("{path}.rs"),
                format!("{path}/mod.rs"),
            ],
        )
    }
}

fn walk_node(
    node: Node<'_>,
    source: &[u8],
    file_path: &str,
    result: &mut ParseResult,
    scope: Scope<'_>,
) {
    match node.kind() {
        "use_declaration" => {
            handle_use(node, source, result);
            return;
        }
        "function_item" => {
            if let Some(name_node) = child_by_field(node, "name")
                .or_else(|| first_named_child_of_kind(node, &["identifier"]))
            {
                let name = qualified(scope.container, &text(source, name_node));
                let kind = if scope.container.is_some() {
                    "method"
                } else {
                    "function"
                };
                result
                    .symbols
                    .push(symbol(source, node, file_path, "rust", name.clone(), kind));
                push_call_edges(node, source, &name, &mut result.edges, &[]);
            }
            return;
        }
        "function_signature_item" => {
            if let Some(container) = scope.container {
                if let Some(name_node) = child_by_field(node, "name")
                    .or_else(|| first_named_child_of_kind(node, &["identifier"]))
                {
                    let name = qualified(Some(container), &text(source, name_node));
                    result
                        .symbols
                        .push(symbol(source, node, file_path, "rust", name, "method"));
                }
            }
            return;
        }
        "struct_item" | "trait_item" => {
            handle_named_class(node, source, file_path, result);
            if node.kind() == "trait_item" {
                let Some(name_node) = child_by_field(node, "name").or_else(|| {
                    first_named_child_of_kind(node, &["type_identifier", "identifier"])
                }) else {
                    return;
                };
                let name = text(source, name_node);
                walk_children_with_scope(node, source, file_path, result, Some(&name));
            }
            return;
        }
        "enum_item" => {
            handle_enum(node, source, file_path, result);
            return;
        }
        "impl_item" => {
            handle_impl(node, source, file_path, result);
            return;
        }
        "type_item" | "const_item" | "static_item" => {
            if let Some(name_node) = child_by_field(node, "name")
                .or_else(|| first_named_child_of_kind(node, &["type_identifier", "identifier"]))
            {
                let name = text(source, name_node);
                result
                    .symbols
                    .push(symbol(source, node, file_path, "rust", name, "variable"));
            }
            return;
        }
        "macro_definition" => {
            if let Some(name_node) = first_named_child_of_kind(node, &["identifier"]) {
                let name = text(source, name_node);
                result
                    .symbols
                    .push(symbol(source, node, file_path, "rust", name, "macro"));
            }
            return;
        }
        _ => {}
    }
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        walk_node(child, source, file_path, result, scope);
    }
}

fn walk_children_with_scope(
    node: Node<'_>,
    source: &[u8],
    file_path: &str,
    result: &mut ParseResult,
    container: Option<&str>,
) {
    let mut cursor = node.walk();
    for child in node.children(&mut cursor) {
        walk_node(
            child,
            source,
            file_path,
            result,
            Scope {
                container,
                in_function: false,
            },
        );
    }
}

fn handle_named_class(node: Node<'_>, source: &[u8], file_path: &str, result: &mut ParseResult) {
    if let Some(name_node) = child_by_field(node, "name")
        .or_else(|| first_named_child_of_kind(node, &["type_identifier", "identifier"]))
    {
        let name = text(source, name_node);
        result
            .symbols
            .push(symbol(source, node, file_path, "rust", name, "class"));
    }
}

fn handle_enum(node: Node<'_>, source: &[u8], file_path: &str, result: &mut ParseResult) {
    let Some(name_node) = child_by_field(node, "name")
        .or_else(|| first_named_child_of_kind(node, &["type_identifier", "identifier"]))
    else {
        return;
    };
    let name = text(source, name_node);
    result.symbols.push(symbol(
        source,
        node,
        file_path,
        "rust",
        name.clone(),
        "class",
    ));
    walk_preorder(node, &mut |variant| {
        if variant.kind() != "enum_variant" {
            return;
        }
        if let Some(variant_name) = first_named_child_of_kind(variant, &["identifier"]) {
            result.symbols.push(symbol(
                source,
                variant,
                file_path,
                "rust",
                qualified(Some(&name), &text(source, variant_name)),
                "variable",
            ));
        }
    });
}

fn handle_impl(node: Node<'_>, source: &[u8], file_path: &str, result: &mut ParseResult) {
    let impl_text = text(source, node);
    let (trait_name, type_name) = impl_header_names(&impl_text);
    if let (Some(trait_name), Some(implementor)) = (trait_name.as_deref(), type_name.as_deref()) {
        result
            .edges
            .push(edge(implementor, trait_name, "implements", None));
        result
            .edges
            .push(edge(trait_name, implementor, "implemented_by", None));
        walk_children_with_scope(node, source, file_path, result, Some(implementor));
        return;
    }
    if impl_header_has_trait_for(&impl_text) {
        let mut types = Vec::new();
        walk_preorder(node, &mut |candidate| {
            if candidate.kind() == "type_identifier" {
                types.push(clean_impl_type_name(&text(source, candidate)));
            }
        });
        if types.len() >= 2 {
            let trait_name = &types[0];
            let implementor = &types[1];
            result
                .edges
                .push(edge(implementor, trait_name, "implements", None));
            result
                .edges
                .push(edge(trait_name, implementor, "implemented_by", None));
            walk_children_with_scope(node, source, file_path, result, Some(implementor));
            return;
        }
    }
    if let Some(type_name) = type_name {
        walk_children_with_scope(node, source, file_path, result, Some(&type_name));
    }
}

fn impl_header_names(impl_text: &str) -> (Option<String>, Option<String>) {
    let Some(header) = impl_text.split('{').next() else {
        return (None, None);
    };
    let rest = header.trim().strip_prefix("impl").unwrap_or(header).trim();
    let rest = strip_leading_impl_generics(rest);
    if let Some((trait_name, implementor)) = rest.split_once(" for ") {
        return (
            nonempty_clean_impl_type_name(trait_name),
            nonempty_clean_impl_type_name(implementor),
        );
    }
    (None, nonempty_clean_impl_type_name(rest))
}

fn impl_header_has_trait_for(impl_text: &str) -> bool {
    let Some(header) = impl_text.split('{').next() else {
        return false;
    };
    let rest = header.trim().strip_prefix("impl").unwrap_or(header).trim();
    strip_leading_impl_generics(rest).contains(" for ")
}

fn strip_leading_impl_generics(value: &str) -> &str {
    let value = value.trim_start();
    if !value.starts_with('<') {
        return value;
    }
    let mut depth = 0_i32;
    for (index, character) in value.char_indices() {
        match character {
            '<' => depth += 1,
            '>' => {
                depth -= 1;
                if depth == 0 {
                    return value[index + character.len_utf8()..].trim_start();
                }
            }
            _ => {}
        }
    }
    value
}

fn nonempty_clean_impl_type_name(value: &str) -> Option<String> {
    let cleaned = clean_impl_type_name(value);
    (!cleaned.is_empty()).then_some(cleaned)
}

fn clean_impl_type_name(value: &str) -> String {
    let mut cleaned = value.trim();
    while let Some(stripped) = cleaned.strip_prefix('&') {
        cleaned = stripped.trim_start();
    }
    cleaned
        .split(" where ")
        .next()
        .unwrap_or(cleaned)
        .split_whitespace()
        .next()
        .unwrap_or(cleaned)
        .split('<')
        .next()
        .unwrap_or(cleaned)
        .rsplit("::")
        .next()
        .unwrap_or(cleaned)
        .trim()
        .to_string()
}

fn handle_use(node: Node<'_>, source: &[u8], result: &mut ParseResult) {
    let Some(argument) = child_by_field(node, "argument") else {
        return;
    };
    handle_use_item(argument, source, "", &mut result.edges);
}

fn handle_use_item(
    node: Node<'_>,
    source: &[u8],
    prefix: &str,
    edges: &mut Vec<crate::models::ParsedEdge>,
) {
    match node.kind() {
        "use_wildcard" => {}
        "scoped_use_list" => {
            let next_prefix = child_by_field(node, "path")
                .map(|path| join_rust_path(prefix, &text(source, path)))
                .unwrap_or_else(|| prefix.to_string());
            if let Some(list) = child_by_field(node, "list") {
                handle_use_item(list, source, &next_prefix, edges);
            }
        }
        "use_list" => {
            let mut cursor = node.walk();
            for child in node.named_children(&mut cursor) {
                handle_use_item(child, source, prefix, edges);
            }
        }
        "use_as_clause" => {
            let Some(path_node) = child_by_field(node, "path") else {
                return;
            };
            let Some(alias_node) = child_by_field(node, "alias") else {
                return;
            };
            let path = text(source, path_node);
            let full_path = join_rust_path(prefix, &path);
            let target = rust_path_leaf(&path).to_string();
            let local = text(source, alias_node);
            edges.push(edge(local, target, "imports", Some(full_path)));
        }
        "scoped_identifier" | "identifier" | "crate" | "self" | "super" => {
            let path = text(source, node);
            let full_path = join_rust_path(prefix, &path);
            let name = rust_path_leaf(&path).to_string();
            edges.push(edge(name.clone(), name, "imports", Some(full_path)));
        }
        _ => {}
    }
}

fn join_rust_path(prefix: &str, path: &str) -> String {
    if prefix.is_empty() {
        path.to_string()
    } else if path.is_empty() {
        prefix.to_string()
    } else {
        format!("{prefix}::{path}")
    }
}

fn rust_path_leaf(path: &str) -> &str {
    path.rsplit("::").next().unwrap_or(path)
}
